tests/cases/compiler/dependentContextualInference1.ts

```diff
@@ -0,0 +1,12 @@
+declare const f:
+  <T extends F<T>>(t: T) => T
+
+type F<T> =
+  { a: unknown
+  , b: (a: T["a" & keyof T]) => unknown
```

Comment:
This test fails if I write `T extends { a: infer X } ? X : never` (ie as it's written in #51377) instead of `T["a" & keyof T]`, which is something I would have not expected as both versions essentially do the same thing.